### PR TITLE
DataIntegrityChecksJob: add OpenTasksWithParentNotOnHold checker

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves finding bad task trees where an open task has a parent task that is not `on_hold`.

### Description
Add `OpenTasksWithParentNotOnHold` checker to `DataIntegrityChecksJob` to find task trees where an open task has a parent task that is not `on_hold`.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Identifies task trees where an open task has a parent task that is not `on_hold`

### Testing Plan
Test code changes in prod:
```ruby
open_tasks_with_parent_not_on_hold = Task.open.joins(:parent).includes(:parent).where.not(parents_tasks: { status: :on_hold });
open_tasks_with_parent_not_on_hold.group(:type, 'parents_tasks.type', 'parents_tasks.status').count
=> {
 ["AodMotionMailTask", "RootTask", "completed"]=>18,
 ["AppealWithdrawalMailTask", "RootTask", "completed"]=>1,
 ["ClearAndUnmistakeableErrorMailTask", "RootTask", "completed"]=>3,
 ["ControlledCorrespondenceMailTask", "RootTask", "cancelled"]=>2,
 ["ControlledCorrespondenceMailTask", "RootTask", "completed"]=>3,
 ["DistributionTask", "RootTask", "completed"]=>2,
 ["EvidenceOrArgumentMailTask", "RootTask", "completed"]=>38,
 ["FoiaColocatedTask", "JudgeAssignTask", "completed"]=>1,
 ["FoiaRequestMailTask", "DistributionTask", "cancelled"]=>1,
 ["FoiaRequestMailTask", "RootTask", "completed"]=>25,
 ["HearingTask", "DistributionTask", "assigned"]=>26,
 ["HearingTask", "DistributionTask", "completed"]=>50,
 ["HearingTask", "RootTask", "assigned"]=>1,
 ["HearingTask", "RootTask", "completed"]=>651,
 ["InformalHearingPresentationTask", "DistributionTask", "completed"]=>13,
 ["InformalHearingPresentationTask", "RootTask", "completed"]=>17,
 ["NoShowHearingTask", "NoShowHearingTask", "completed"]=>247,
 ["OtherMotionMailTask", "RootTask", "completed"]=>7,
 ["PrivacyActRequestMailTask", "RootTask", "completed"]=>1,
 ["PrivacyActTask", "AddressVerificationColocatedTask", "completed"]=>1,
 ["PrivacyActTask", "ExtensionColocatedTask", "completed"]=>4,
 ["PrivacyActTask", "IhpColocatedTask", "completed"]=>1,
 ["PrivacyActTask", "OtherColocatedTask", "cancelled"]=>3,
 ["PrivacyActTask", "OtherColocatedTask", "completed"]=>9,
 ["PrivacyActTask", "PreRoutingFoiaColocatedTask", "completed"]=>2,
 ["ReconsiderationMotionMailTask", "RootTask", "cancelled"]=>1,
 ["ReconsiderationMotionMailTask", "RootTask", "completed"]=>131,
 ["ReturnedUndeliverableCorrespondenceMailTask", "RootTask", "completed"]=>2,
 ["StatusInquiryMailTask", "RootTask", "cancelled"]=>3,
 ["StatusInquiryMailTask", "RootTask", "completed"]=>36,
 ["TrackVeteranTask", "RootTask", "cancelled"]=>2,
 ["TrackVeteranTask", "RootTask", "completed"]=>260,
 ["VacateMotionMailTask", "RootTask", "completed"]=>11,
 ["VeteranRecordRequest", "RootTask", "completed"]=>2}

IGNORED_TASKS_WITH_CLOSED_ROOTTASK_PARENT = [
  *MailTask.descendants.map(&:name),
  "TrackVeteranTask",
  "FoiaTask"
].freeze

def ignored_tasks_with_closed_root_task_parent
  Task.open.joins(:parent).includes(:parent).where(type: IGNORED_TASKS_WITH_CLOSED_ROOTTASK_PARENT)
    .where(parents_tasks: { type: "RootTask", status: %w[completed cancelled] })
end

task_ids = open_tasks_with_parent_not_on_hold.pluck(:id) - ignored_tasks_with_closed_root_task_parent.pluck(:id);
tasks_with_parents = Task.where(id: task_ids).joins(:parent).includes(:parent);
grouped_suspect_tasks = tasks_with_parents.group(:appeal_type, :type, "parents_tasks.type", "parents_tasks.status").count
=> {
 ["Appeal", "DistributionTask", "RootTask", "completed"]=>2,
 ["Appeal", "FoiaColocatedTask", "JudgeAssignTask", "completed"]=>1,
 ["Appeal", "HearingTask", "DistributionTask", "assigned"]=>26,
 ["Appeal", "HearingTask", "DistributionTask", "completed"]=>48,
 ["Appeal", "HearingTask", "RootTask", "completed"]=>1,
 ["Appeal", "InformalHearingPresentationTask", "DistributionTask", "completed"]=>13,
 ["Appeal", "InformalHearingPresentationTask", "RootTask", "completed"]=>17,
 ["Appeal", "VeteranRecordRequest", "RootTask", "completed"]=>1,
 ["LegacyAppeal", "HearingTask", "RootTask", "assigned"]=>1,
 ["LegacyAppeal", "HearingTask", "RootTask", "completed"]=>651,
 ["LegacyAppeal", "NoShowHearingTask", "NoShowHearingTask", "completed"]=>247,
 ["LegacyAppeal", "PrivacyActTask", "AddressVerificationColocatedTask", "completed"]=>1,
 ["LegacyAppeal", "PrivacyActTask", "ExtensionColocatedTask", "completed"]=>4,
 ["LegacyAppeal", "PrivacyActTask", "IhpColocatedTask", "completed"]=>1,
 ["LegacyAppeal", "PrivacyActTask", "OtherColocatedTask", "cancelled"]=>3,
 ["LegacyAppeal", "PrivacyActTask", "OtherColocatedTask", "completed"]=>9,
 ["LegacyAppeal", "PrivacyActTask", "PreRoutingFoiaColocatedTask", "completed"]=>2
}
```

Investigate:
```ruby
dts=DistributionTask.open.joins(:parent).includes(:parent).where(parents_tasks: { type: "RootTask", status: :completed });
puts dts.map{|t| t.appeal.tree}

dts.map{|t| [t.appeal.id, t.appeal.uuid]} # Use this to search in Slack for previous investigations
=> [[30321, "25c8df2e-e6a8-4935-88f9-f88e9ae41552"], [170975, "d9fe18c6-f39e-4ea0-9614-ca5e388c5e02"]]
```

### CODE REVIEW

File: app/services/open_tasks_with_closed_at_checker.rb [hunk 1]
```diff
@@ -28,17 +19,4 @@ def slack_channel
...
-  end
-
-  # It's acceptable to have a closed RootTask parent for:
-  # - MailTasks -- https://dsva.slack.com/archives/CJL810329/p1634239591067100?thread_ts=1634224678.055200&cid=CJL810329
-  # - TrackVeteranTask -- https://dsva.slack.com/archives/CJL810329/p1634581182080100?thread_ts=1634553075.073600&cid=CJL810329
-  IGNORED_TASKS_WITH_CLOSED_ROOTTASK_PARENT = MailTask.descendants.map(&:name) + ["TrackVeteranTask"]
-  def ignored_tasks_with_closed_root_task_parent
-    Task.open.joins(:parent).includes(:parent).where(type: IGNORED_TASKS_WITH_CLOSED_ROOTTASK_PARENT)
-      .where(parents_tasks: { type: "RootTask", status: %w[completed cancelled] })
-  end
```
Comment (yoomlam) [inline app/services/open_tasks_with_closed_at_checker.rb:43]: This scenario is covered in the new `OpenTasksWithParentNotOnHold` class, so all this code is moved over there.

File: app/services/open_tasks_with_closed_at_checker.rb [hunk 2]
```diff
@@ -3,20 +3,11 @@
...
+    return if suspect_tasks.empty?
 
-    suspect_task_ids = open_tasks_with_closed_parent.pluck(:id) - ignored_tasks_with_closed_root_task_parent.pluck(:id)
-    if suspect_task_ids.count > 0
-      add_to_report "#{suspect_task_ids.count} open " +
-                    "Task".pluralize(suspect_task_ids.count) +
-                    " with a closed parent Task"
-      add_to_report "Counts: " + Task.where(id: suspect_task_ids).group(:type).count.to_s
-      add_to_report "Verify with `Task.where(id: [#{suspect_task_ids.join(',')}])`"
-    end
```
Comment (yoomlam) [inline app/services/open_tasks_with_closed_at_checker.rb:19]: This scenario is covered in the new `OpenTasksWithParentNotOnHold` class, so all this code is moved over there.

File: app/services/open_tasks_with_parent_not_on_hold.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  # - TrackVeteranTask -- https://dsva.slack.com/archives/CJL810329/p1634581182080100?thread_ts=1634553075.073600&cid=CJL810329
+  IGNORED_TASKS_WITH_CLOSED_ROOTTASK_PARENT = [
+    *MailTask.descendants.map(&:name),
+    "TrackVeteranTask"
+  ].freeze
+
+  def ignored_tasks_with_closed_root_task_parent
+    Task.open.joins(:parent).includes(:parent).where(type: IGNORED_TASKS_WITH_CLOSED_ROOTTASK_PARENT)
+      .where(parents_tasks: { type: "RootTask", status: %w[completed cancelled] })
+  end
```
Comment (yoomlam) [inline app/services/open_tasks_with_parent_not_on_hold.rb:49]: Moved from `OpenTasksWithClosedAtChecker`

File: spec/services/open_tasks_with_closed_at_checker_spec.rb [hunk 4]
```diff
@@ -12,61 +12,12 @@
...
   end
 
-  let!(:open_task_with_closed_parent) do
-    appeal = create(:appeal)
-    parent = create(:task, appeal: appeal)
-    task = create(:task, :assigned, parent: parent)
-    parent.update!(closed_at: Time.zone.now, status: :completed)
-    task
-  end
-
```
Comment (yoomlam) [inline spec/services/open_tasks_with_closed_at_checker_spec.rb:22]: This scenario is covered in the new test for the `OpenTasksWithParentNotOnHold` class, so all this code is moved over there.

File: app/services/open_tasks_with_closed_at_checker.rb [hunk 5]
```diff
@@ -3,20 +3,11 @@
 class OpenTasksWithClosedAtChecker < DataIntegrityChecker
   def call
     suspect_tasks = open_tasks_with_closed_at_defined
-    if suspect_tasks.count > 0
-      add_to_report "#{suspect_tasks.count} open " + "Task".pluralize(suspect_tasks.count) + " with a closed_at value"
-      add_to_report "Verify data error with `Task.open.where.not(closed_at: nil)`"
-      add_to_report "These tasks likely were manually re-opened and should have closed_at set to NULL"
-    end
+    return if suspect_tasks.empty?
```
Comment (pbradin) [inline app/services/open_tasks_with_closed_at_checker.rb:6]: surprised this wasn't already there; great addition.

**Review by pbradin — APPROVED** [review]
I like the changes, all looks good and is working well.

**Pull request merged**